engine/repositories/thread_fs_repository.h

```diff
@@ -0,0 +1,62 @@
+#pragma once
+
+#include <filesystem>
+#include <shared_mutex>
+#include <unordered_map>
+#include "common/repository/thread_repository.h"
+#include "common/thread.h"
+#include "utils/logging_utils.h"
+
+class ThreadFsRepository : public ThreadRepository {
+ private:
+  constexpr static auto kThreadFileName = "thread.json";
+  constexpr static auto kThreadContainerFolderName = "threads";
+
+  mutable std::shared_mutex map_mutex_;
+  mutable std::unordered_map<std::string, std::unique_ptr<std::shared_mutex>>
+      thread_mutexes_;
+
+  /**
+   * The path to the data folder.
+   */
+  std::filesystem::path data_folder_path_;
+
+  std::shared_mutex& GrabThreadMutex(const std::string& thread_id) const;
+
+  std::filesystem::path GetThreadPath(const std::string& thread_id) const;
+
+  /**
+   * Read the thread file and parse to Thread from the file system.
+   */
+  cpp::result<OpenAi::Thread, std::string> LoadThread(
+      const std::string& thread_id) const;
+
+  cpp::result<void, std::string> SaveThread(OpenAi::Thread& thread);
```

Comment:
Seems like can use const& for thread parameter?

Reply:
because in the implementation we will call a function `.ToJson`. In order to make it as const, we will need to make `ToJson` const which introduce a lot of changes. I will update it later on.